### PR TITLE
feat(server): allow customize server external url

### LINKED ISSUE #7252: email url ​ can not be set to contain port
### What happened?

I installed the self-host version with docker, I want to know how to set the BaseUrl which is used in the mailer. How to Set the BaseUrl with port (Ex: http://192.168.1.1:3010 ) OR can I try to define the 'AFFINE_SERVER_HOST' in '.env' with port?
```
AFFINE_SERVER_HOST=http://192.168.1.1:3010
```

### Distribution version

Linux

### What browsers are you seeing the problem on if you're using web version?

_No response_

### Are you self-hosting?

- [X] Yes

### Relevant log output

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by forehalo**
closes https://github.com/toeverything/AFFiNE/issues/7252

### CODE REVIEW

**Comment by forehalo** [comment]
* **#7270** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/7270?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `canary`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @forehalo and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: packages/backend/server/src/fundamentals/helpers/url.ts [hunk 1]
```diff
@@ -6,19 +8,38 @@ import { Config } from '../config';
...
+          'Invalid `server.externalUrl` configured. It must be a valid url.'
+        );
+      }
+
+      const externalUrl = new URL(this.config.server.externalUrl);
+
+      this.origin = externalUrl.origin;
+      this.baseUrl =
+        externalUrl.origin + externalUrl.pathname.replace(/\/$/, '');
+      console.log(externalUrl);
```
Comment (darkskygit) [inline packages/backend/server/src/fundamentals/helpers/url.ts]: cleanup logs

File: packages/backend/server/src/fundamentals/helpers/url.ts [hunk 2]
```diff
@@ -6,19 +8,38 @@ import { Config } from '../config';
...
+    } else {
+      this.origin = [
+        this.config.server.https ? 'https' : 'http',
+        '://',
+        this.config.server.host,
```
Comment (darkskygit) [inline packages/backend/server/src/fundamentals/helpers/url.ts:33]: hostname need to be check that only included valid chars

**Review by darkskygit — APPROVED** [review]

File: packages/backend/server/src/fundamentals/helpers/url.ts (hunk 2 above)
Comment (forehalo) [reply packages/backend/server/src/fundamentals/helpers/url.ts:33]: owners should know what it should be

**Pull request merged**